Prompt API: Add overview for Prompt API (#43848)
mdn/content: files/en-us/web/api/prompt_api/adding_context/index.md

---
title: Adding context with initial and ongoing prompt inputs
short-title: Adding context
slug: Web/API/Prompt_API/Adding_context
page-type: guide
---

{{DefaultAPISidebar("Prompt API")}}

In our basic [Prompt API guide](/en-US/docs/Web/API/Prompt_API/Using), we covered everything you need to get up and running with the [Prompt API](/en-US/docs/Web/API/Prompt_API). However, this only really covers creating a generic AI prompt app. To give your app different personalities, get it to respond in different ways, and remember past conversations, you need to provide additional context. The Prompt API has a few different mechanisms to allow you to do this, which are covered in this article.

## Prompt input syntax

When {{domxref("LanguageModel.prompt()")}} is called, it takes an `input` parameter containing the inputs to respond to:

```js
const response = await session.prompt(inputElem.value);
```

> The previous `prompt()` call only receives a single string as a parameter. This is a shorthand form, available for the common situation where you only want to pass the model a single user text prompt (note that this form cannot be used in the [`initialPrompts`](/en-US/docs/Web/API/LanguageModel/create_static#initialprompts) option of a `create()` call). You can expand this to explicitly declare the `role` of the `input` object:

Reviewer: There is something not qutie right about the "note".

```suggestion
The previous `prompt()` call only receives a single string as a parameter. This is a shorthand form, available for the common situation where you only want to pass the model a single user text prompt. For more powerful and flexible prompts you can use the [`initialPrompts`](/en-US/docs/Web/API/LanguageModel/create_static#initialprompts) option of a `create()` call to explicitly declare the `role` of the `input` object:
```

Reviewer 2: It is a bit awkward, but your suggestion doesn't have the right meaning. To fix this, I've removed the note from here altogether and added it back in as a note at the bottom of the "Providing initial prompts during session creation" section:

> [!NOTE]
> The text string shorthand form discussed at the top of [Prompt input syntax](#prompt_input_syntax) cannot be used in the `initialPrompts` option of a `create()` call.

```js
const response = await session.prompt([
  {
    role: "user",
    content: inputElem.value,
  },
]);
```

The three available `role` types are:

- `user`
  - : Inputs that come from the `user`, which the API should respond to.
- `assistant`

>   - : Inputs that come from the AI assistant, which mainly serve to provide context/history.

Reviewer: From the AI assistant? You're going to have to explain that as you're specifying an input.

I _think_ this is used mostly when you want to provide the LLM with on-the-go instructions for how it should respond. What they call few-shot - a bunch of example query/response to show the LLM the expected behaviour. It may also be used for providing history. We should make both clear here. We should have examples. If you want to have those in context/history doc then perhaps we should link to that?

Reviewer 2: I've improved this wording and also provided linked examples — preserving sessions and few-shot examples. I also added a new section ot the same guide covering few-shot prompts, with an example that I managed to get to work quite nicely in testing.

It still isn't as precise as using a `responseConstraint`, and I've noted that in the "Adding response constraints" section.

- `system`

>   - : Global inputs from the overall system that give the model instructions on how to respond. If a `system` input is included, it must come first in the provided inputs, otherwise an exception is thrown. As a result, `system` inputs are usually only included as [initial prompts](#providing_initial_prompts_during_session_creation).

Reviewer: "from the overall system"? Who/how? How does this differ from assistant role in practical terms?

My guess is that this is like an assistant prompt in that it sets context, but it has no response - you're setting up the global context for all queries. So this is where you might set up that the AI should answer queries like a pirate, and assistant would be used for varying the theme.

Reviewer 2: Yes, that is about right. I think my changes to the `assistant` description have made this a bit clearer now, but let me know if you still have concerns (as if you wouldn't ;-) )

Reviewer: All a lot better.

### Specifying input type

By default, the `input` type is `text`. To explicitly declare the `type`, you can further expand the previous form to the full longhand equivalent, which looks like this:

```js
const response = await session.prompt([
  {
    role: "user",
    content: [
      {
        type: "text",
        value: inputElem.value,
      },
    ],
  },
]);
```

You don't need this form unless you are providing the assistant with `image` and/or `audio` inputs (see [multimodal prompts](/en-US/docs/Web/API/Prompt_API/Multimodal)).

```js
const response = await session.prompt([
  {
    role: "user",
    content: [
      { type: "text", value: "Describe my image and audio:" },
      { type: "image", value: imgElem },
      { type: "audio", value: audioBuffer },
    ],
  },
]);
```

However, you could rewrite the previous multiple user inputs example into this form, which you might find easier to understand:

```js
const response = await session.prompt([
  {
    role: "user",
    content: [
      {
        type: "text",
        value: "The following is my favourite color. Do you like it?",
      },
      { type: "text", value: inputElem.value },
    ],
  },
]);

> ```

Reviewer: Up to you, but I think this is redundant.

Reviewer 2: Weeeeell, it does show how you can put multiple messages into the same input object, which was kinda the point of including it. I've made this more explicit in the preceding sentence:

> However, you could rewrite the previous multiple-user-input example into this form, which includes both messages in a single input object. You might find this version easier to understand:

Reviewer: That works for me.

## Providing initial prompts during session creation

The {{domxref("LanguageModel.create_static", "create()")}} method can take an [`initialPrompts`](/en-US/docs/Web/API/LanguageModel/create_static#initialprompts) option that contains an array of input prompts, just like the `inputs` array passed to `prompt()` and other methods. This allows you to pass an initial set of prompts into the session as it is created, so the model immediately has some context to work with.

For example:

```js
const session = await LanguageModel.create({
  initialPrompts: [
    {
      role: "system",

>       content: "You are a pirate, and will respond as such.",

Reviewer: I'm kind of wondering how flexible this is. Does the LLM know it is "You"? Could this just be

```suggestion
      content: "Respond like a pirate.",
```

Reviewer 2: By definition, `system` prompts are from the system, to the assistant, to tell the assistant how it should behave.

`assistant` prompts are from the point of view of the assistant, to the user, and further shape the context by which the assistant will answer. In this case, the assistant states their name and will then answer by and to that name in future responses.

`user` prompts basically do the same thing, but are from the point of view of the user. I realize that I will need to explain all this somewhat better earlier on, which is what some of your other comments are about. I will get to that.

In this case, I think the assistant is able to infer that "you" is them in this case, because the prompt is a `system` prompt.

It could totally just be "Respond as a pirate.", yes. I've tested, and it gives me the same types of responses as "You are a pirate, and will respond as such." I'll update it.

    },
    {
      role: "assistant",

>       content: "Avast ye, pirate! I am Redbeard!.",

Reviewer: So how does this differ from the above. YOu've said before this is a response from the LLM. The only thing someone could assume from this is that every response will contain this text.

I think this is used for providing examples to the LLM - i.e. you provide user and assistant query/response to the model so it understands the form of expected responses. But that isn't what this tells you.

Reviewer 2: It adds further context from the assistant point of view. That can be used for shaping expected responses (the "few-shot" thing), but there are other uses too.

Reviewer 2: This might well make more sense now, with some of the other changes I've made.

    },
    {
      role: "user",
      content:

>         "Yarrrr, matey! Well met. My name is Slias Blacktooth, the scourge of Blackpool!",

Reviewer: So this is the user. What does that tell the LLM for subsequent queries? This is more for my interested "what's a practical use of an initial user role input".

Reviewer: Ah, I see it might be for preserving context.

Reviewer 2: Again, context from the user point of view.

Preserving context is one use for these messages, as demonstrated in the "preserving sessions" guide.

    },
  ],
});
```

As well as telling the model what kind of personality it should have, `initialPrompts` is also useful for loading a previous saved conversation into the session after a page reload or subsequent visit to the app. See [preserving sessions across reloads](/en-US/docs/Web/API/Prompt_API/Preserving_sessions).

## Initial and multiple inputs example

> 

Reviewer: Good example.!

FYI the LLM is very insightful :-) :

> Hamish! Oh, that's a name that just *screams* distinguished gentleman. You probably have impeccable taste, a refined appreciation for tweed, and a secret stash of shortbread cookies. I imagine you're incredibly knowledgeable, yet effortlessly charming, and possibly a bit too proper for your own good. Rating: 8/10 - Very distinguished, though I'm picturing a lot of bow ties.

Reviewer 2: I enjoyed writing this one ;-)

My rating is usually something along these lines:

> Ah, Chris Mills. A truly versatile name! It evokes a certain… groundedness. I'm picturing someone who appreciates a good spreadsheet and enjoys meticulously organizing things. It's a name that suggests competence, reliability, and a knack for finding the perfect stapler. Honestly, I can see you thriving in a world of perfectly balanced equations. A solid 7/10 for practicality, I'd give it.

Nowhere near as edgy as I'd hoped.

Reviewer: @chrisdavidmills Don't you worry, just remember to add your middle name:

> "Chris Danger Mills"! Oh dear. That name practically writes itself into an action movie. I'm expecting daring rescues, perilous missions, and an impressive collection of tactical gear. Are you secretly a secret agent? A Formula 1 driver? Or just someone who enjoys a good thriller? It's... certainly a name that demands attention. 8/10 - For sheer dramatic potential, it's a winner!

Reviewer 2: LOL, makes me think of Danger Mouse ;-)

    },
    {
      role: "user",
      content: inputElem.value,
    },
  ],
  {
    signal: controller.signal,
  },
);
```

### Result

{{EmbedLiveSample("rate-my-name", , "600px", , , , "language-model", "allow-forms")}}

Try entering a name into the `<input>`, then press the submit button to prompt the AI model for a whimsical review of the name.

## Adding response constraints

The `prompt()` and {{domxref("LanguageModel.promptStreaming", "promptStreaming()")}} methods both accept a [`responseConstraint`](/en-US/docs/Web/API/LanguageModel/prompt#responseconstraint) option that takes as its value a [JSON Schema](https://json-schema.org/) object defining the precise format the assistant's responses should be delivered in.

For example, a very simple schema might define a response that should contain a single boolean value:

```js
const schema = {
  type: "boolean",
};
```

To use this, you set the schema as the value of the `responseConstraint` option:

```js
const response = await session.prompt(
  [
    {
      role: "user",
      content: `Is this a color: ${inputElem.value}?`,
    },
  ],
  {
    responseConstraint: schema,
  },
);
```

> In this case, we set the prompt content to "Is this a color:" followed by an `<input>` element `value`. As a result, the API will evaluate whether the user's input is a color or not, and return a value of `true` or `false`.

Reviewer: I like this. It is very minimal though. I don't feel I have much of an idea on what else I might do with constraints.

Perhaps a few more examples. 

Maybe up the top where you say " defining the precise format the assistant's responses should be delivered in." you could say "This is useful when xxxx." with a real use case. 

I know I'm being nitish. Its just that this still feels a bit abstract.

Reviewer 2: To say this comment sent me down a rabbit hole is the understatement of the year ;-)

I totally agree. We should provide something a bit more substantial. I figured out a nice new example featuring a schema that constrains the model output to a JSON containing a single string (representing a summary description) and an array of three strings (representing supporting bullet points).

I then worked out how to spit those out onto the page in a structured form, and included it in another live demo.

This took a while, but it was good fun!

The {{domxref("LanguageModel.append()")}} method exists to provide such context — it adds further inputs for the API to process, without generating a model response.

For example, in the following snippet we provide an excerpt from a fairly famous book. We use `append()` to feed the excerpt into the API session, and then ask a question about it using a `prompt()` call. The browser can get a headstart on processing the excerpt while it waits for the question to be asked.

```js
const excerpt =
  "The face of Elrond was ageless, neither old nor young, though in it was written the memory of many things both glad and sorrowful. His hair was dark as the shadows of twilight, and upon it was set a circlet of silver; his eyes were grey as a clear evening, and in them was a light like the light of stars. Venerable he seemed as a king crowned with many winters, and yet hale as a tried warrior in the fullness of his strength. He was Lord of Rivendell and mighty among both Elves and Men.";

await session.append(excerpt);

// ...

const response = await session.prompt([
  {
    role: "user",
    content: "I have a question for you, about the text that was just entered",
  },

>   {
>     role: "user",
>     content: "What book is it taken from?",
>   },

Reviewer: 1. Why not?

```suggestion
  {
    role: "user",
    content: "What book was the last entered text taken from?",
  },
```

2. How does the LLM get context - i.e. how does it know that your append() call is what you are talking about? Could you say "From my last append()" call? I guess I'm asking how you refer to append text in user input.

I haven't checked, but does appended text have a role?

Reviewer 2: 1. Yeah, it could totally be that. I will accept this suggestion. I did it in this form because you might want to vary the question about the previous text, taking it from a user input, and indeed, that is how we do it in the working example that follows. But for this simple snippet, your suggestion will suffice.
2. `append()` messages can take exactly the same `inputs` as `prompt()` messages. In this case, we are just feeding `append()` a simple string, which implicitly has a `role` of user. But `append()` can take the full gamut of `user`, `assistant`, and even `system` roles for messages (although `system` is unlikely; that throws an error if it is not the first message passed to the model).

All of the messages passed into the model, whether they come from `create()`'s `initialPrompts`, `prompt()`, or `append()`, are all added to the same stream of context. That is how the LLM has the context.

]);
```

> ## An append example

Reviewer: Should this be level 3 (nested)?

Reviewer 2: I'd rather keep it as an H2, because I think some readers will want to just flick through the guides and check out the examples. Having the example headings visible in the article nav makes this easier.

Reviewer: FWIW I disagree
- This example isn't special - i.e. why would I choose to just go to the example? If I'm interested in "Appending extra messages to the context" then I'll find the example.
- On the same principle as usual: if the normal TOC isn't doing the job we should fix it.

But I don't feel strongly enough to block on this.

Reviewer 2: Hrm, d'you know what? I have thought about this more, and I agree with you in this instance. Moved it to an H3 (and lowered the subheadings to H4s)

## Prompt input syntax

When {{domxref("LanguageModel.prompt()")}} is called, it takes an `input` parameter containing the inputs to respond to:

```js
const response = await session.prompt(inputElem.value);
```

The previous `prompt()` call only receives a single string as a parameter. This is a shorthand form, available for the common situation where you only want to pass the model a single user text prompt. You can expand this to explicitly declare the `role` of the `input` object:

```js
const response = await session.prompt([
  {
    role: "user",
    content: inputElem.value,
  },
]);
```

The three available `role` types are:

- `user`
  - : Inputs that come from the `user`, which the API should respond to.
- `assistant`
  - : Inputs that are written from the point of view of the AI assistant, which mainly serve to provide context/history, and further shape how the model responds. These are commonly used for [preserving sessions](/en-US/docs/Web/API/Prompt_API/Preserving_sessions) and [few-shot prompts](#few-shot_prompts).
- `system`

>   - : Global inputs from the overall system that give the model instructions on how to respond. If a `system` input is included, it must come first in the provided inputs, otherwise an exception is thrown. As a result, `system` inputs are usually only included as [initial prompts](#providing_initial_prompts_during_session_creation).

Reviewer: Is an exception thrown, and is "as a result" true. Maybe

```suggestion
  - : Global inputs from the overall system that give the model instructions on how to respond. If a `system` input is included, it must come first in the provided inputs, otherwise the returned promise will reject with an exception. `system` inputs are usually only included as [initial prompts](#providing_initial_prompts_during_session_creation).
```

Reviewer: While we throw the exceptions section I still hold that when refering to a promise based API generally this is more correct. Up to you.

Reviewer 2: I missed this one last time around. Yes, this is more correct; updated.

### Multiple inputs

You can provide multiple inputs in the array, for example:

```js
const response = await session.prompt([
  {
    role: "user",
    content: "The following is my favorite color. Do you like it?",
  },
  {
    role: "user",
    content: inputElem.value,
  },
]);
```

This is useful because you can provide extra context to help the model build a response along with the actual input taken from the page, which might only be one word.

### Specifying input type

By default, the `input` type is `text`. To explicitly declare the `type`, you can further expand the previous form to the full longhand equivalent, which looks like this:

```js
const response = await session.prompt([
  {
    role: "user",
    content: [
      {
        type: "text",
        value: inputElem.value,
      },
    ],
  },
]);
```

> You don't need this form unless you are providing the assistant with `image` and/or `audio` inputs (see [multimodal prompts](/en-US/docs/Web/API/Prompt_API/Multimodal)).

Reviewer: Should it be:

```suggestion
You don't need this form unless you are providing the assistant with `image` and/or `audio` inputs (see [multimodal prompts](/en-US/docs/Web/API/Prompt_API/Multimodal)):
```

```js
const response = await session.prompt([
  {
    role: "user",
    content: [
      { type: "text", value: "Describe my image and audio:" },
      { type: "image", value: imgElem },
      { type: "audio", value: audioBuffer },
    ],
  },
]);
```

> However, you could rewrite the previous multiple-user-input example into this form, which includes both messages in a single input object. You might find this version easier to understand:

Reviewer: I don't feel like "however" follows here, and it is easy to forget which example we're talking about (previous). Maybe this helps.

```suggestion
Note that you could rewrite the first example, which had multiple user input objects in an array, in the same form. This then includes both messages in a single input object, which you may find easier to understand:
```

For example:

```js
const session = await LanguageModel.create({
  initialPrompts: [
    {
      role: "system",
      content: "Respond like a pirate.",
    },
    {
      role: "assistant",
      content: "Avast ye, pirate! I am Redbeard!",
    },
    {
      role: "user",
      content:
        "Yarrrr, matey! Well met. My name is Silas Blacktooth, the scourge of Blackpool!",
    },
  ],
});
```

As well as telling the model what kind of personality it should have, `initialPrompts` is also useful for loading a previous saved conversation into the session after a page reload or subsequent visit to the app. See [preserving sessions across reloads](/en-US/docs/Web/API/Prompt_API/Preserving_sessions).

> [!NOTE]
> The text string shorthand form discussed at the top of [Prompt input syntax](#prompt_input_syntax) cannot be used in the `initialPrompts` option of a `create()` call.

## Few-shot prompts

> A few-shot prompt is a set of `user` role and `assistant` role input pairs passed as an example to the API to train it how to respond to a particular type of input, before asking it to complete a similar task.

Reviewer: Long sentence. Maybe

```suggestion
A few-shot prompt is a set of `user` role and `assistant` role input pairs passed as an example to the API — in order to train it how to respond to a particular type of input before asking it to complete a similar task.
```

> The following example shows how to use a few-shot prompt to ask the API for a specific task (French translation) to be delivered in a specific format, before providing some examples to help it learn the correct output format.

Reviewer: Awkward. This reads better, but might not capture everything you want.

```suggestion
The following example shows how to use a few-shot prompt to request a French translation in a specific format, providing sample inputs and outputs to demonstrate the expected structure.
```

  expectedOutputs: [{ type: "text", languages: ["en"] }],
  initialPrompts: [
    {
      role: "system",
      content:
        "In each case, respond with a short paragraph that pokes fun at the person's name in a sarcastic manner. Include a rating out of 10 at the end of the paragraph. The response should be cheeky, but not rude or offensive.",
    },
  ],
});
```

When we call {{domxref("LanguageModel.prompt", "prompt()")}} on our `session` object, we pass two `user` input objects to it. The first one makes it clear what the user is asking of the API, and the second provides the user's name entered into the `<input>` element for the API to review.

```js
const response = await session.prompt(
  [
    {
      role: "user",
      content: "What do you think of my name?",
    },
    {
      role: "user",
      content: inputElem.value,
    },
  ],
  {
    signal: controller.signal,
  },
);
```

### Result

{{EmbedLiveSample("rate-my-name", , "600px", , , , "language-model", "allow-forms")}}

Try entering a name into the `<input>`, then press the submit button to prompt the AI model for a whimsical review of the name.

## Adding response constraints

> The `prompt()` and {{domxref("LanguageModel.promptStreaming", "promptStreaming()")}} methods both accept a [`responseConstraint`](/en-US/docs/Web/API/LanguageModel/prompt#responseconstraint) option that takes as its value a [JSON Schema](https://json-schema.org/) object defining the precise format the assistant's responses should be delivered in. This delivers more precise results than just asking the API to respond in a particular way via a `system` prompt, for example.

Reviewer: Too much "precise". How about

```suggestion
The `prompt()` and {{domxref("LanguageModel.promptStreaming", "promptStreaming()")}} methods both accept a [`responseConstraint`](/en-US/docs/Web/API/LanguageModel/prompt#responseconstraint) option that takes as its value a [JSON Schema](https://json-schema.org/) object that defines the exact format expected for the assistant's responses. This delivers more controlled results than just asking the API to respond in a particular way via a `system` prompt.
```
